### ISSUE TITLE
`--write` does not respect yamllint config  requiring quotes

### ISSUE THREAD

**Issue by jcgruenhage** [ansible/ansible-lint#3072]
##### Summary
Our yamllint config requires quotes, like this:

```yaml
  quoted-strings:
    quote-type: "double"
    allow-quoted-quotes: true
    required: true
```

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console (paste below)
$ ansible-lint --version
ansible-lint 6.13.1 using ansible 2.14.2
```

- ansible installation method: OS package
- ansible-lint installation method: OS package

##### STEPS TO REPRODUCE

- Write ansible code that quotes all strings
- Run yamllint, see that it's happy
- Run `ansible-lint --write`
- Ryn yamllint again, see that it's not happy anymore

##### Desired Behavior

`ansible-lint --write` should add quotes when the yamllint config says they are required

##### Actual Behavior

`ansible-lint --write` removes quotes, even though the yamllint config says they are required

##### Minimal example

```console
test on  main took 2s 
❯ git --no-pager show HEAD
commit 27f8ace5fd8446cfa64dde29d46c967149d3e813 (HEAD -> main)
Author: Jan Christian Grünhage <jan.christian@gruenhage.xyz>
Date:   Mon Feb 20 17:55:00 2023 +0100

    initial commit

diff --git a/.yamllint.yml b/.yamllint.yml
new file mode 100644
index 0000000..330cc32
--- /dev/null
+++ b/.yamllint.yml
@@ -0,0 +1,5 @@
+---
+rules:
+  quoted-strings:
+    quote-type: "double"
+    allow-quoted-quotes: true
diff --git a/playbook.yml b/playbook.yml
new file mode 100644
index 0000000..0c8712b
--- /dev/null
+++ b/playbook.yml
@@ -0,0 +1,3 @@
+---
+- name: "I'm a test play!"
+  hosts: "all"

test on  main 
❯ yamllint playbook.yml

test on  main 
❯ ansible-lint --write playbook.yml

Modified 1 files.
Passed with production profile: 0 failure(s), 0 warning(s) on 1 files.

test on  main [!] took 2s 
❯ git --no-pager diff     
diff --git a/playbook.yml b/playbook.yml
index 0c8712b..641205f 100644
--- a/playbook.yml
+++ b/playbook.yml
@@ -1,3 +1,3 @@
 ---
-- name: "I'm a test play!"
-  hosts: "all"
+- name: I'm a test play!
+  hosts: all

test on  main [!] 
❯ yamllint playbook.yml
playbook.yml
  2:9       error    string value is not quoted with double quotes  (quoted-strings)
  3:10      error    string value is not quoted with double quotes  (quoted-strings)
```

**Comment by jcgruenhage** [comment]
Ah, right, reasons I consider this a bug:
 - https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/yaml_utils.py#L84-L108 explicitly loads the yamllint config and checks what quote style is preferred, so it should go the extra step to also check whether quotes are supposed to be used or not.
 - the minimal example posted above passes ansible-lint without errors before using `--write`, and produces errors after using `--write`.

**Comment by MarkusTeufelberger** [comment]
As far as I know, the `--write` mode will only write a certain style of YAML (that should be relatively close to https://prettier.io/). I don't think that every possible style that's covered by yamllint or the YAML spec will be supported. In the code that you quote it even states "if we detect local yamllint config we use it but raise a warning as this is likely to get out of sync with our internal config."

**Comment by ssbarnea** [comment]
We do not plan to support any custom yamllint configuration for the writer but we will keep this bug open because we need two fixes related to the writer:

- ensure we quote octals correctly
- display a warning when we reformat if the user has a custom yamllint config file, one that states that the writer is not configurable

**Comment by jcgruenhage** [comment]
Could we change the second bit to *not* reformat yaml if there is a custom yamllint config? Using ansible-lint --write for other stuff in the future (like replacing short names with fqcns or whatever, that'd be really nice, but ideally without breaking the yaml stuff that we have custom configs for.

(The big thing where we differ is quoting all strings because of the several footguns that yaml has where unquoted stuff is *maybe* a string and *maybe* something else)

**Comment by MarkusTeufelberger** [comment]
I'd rather just have ansible-lint error out on `--write` if any custom non-empty yamllint config exists or have the individual rewriting steps be individually callable (`--write-booleans`, `--write-quotes` etc.).

**Comment by jcgruenhage** [comment]
But that means that I can't use `--write` to do everything that I want at once? Maybe a warning and skipping that bit or some form of configuration for that could be something we can agree on?

**Comment by ssbarnea** [comment]
@MarkusTeufelberger @jcgruenhage Let me put it this way, is not that we do not want to allow users to configure how it works, is that realistically we don't have the resources to implement and maintain these features, especially as we fully rely on 3rd party ruamel.yaml library for that, which has known open bugs that prevent a proper rewrite of the files. Is not our own code that does the dumping, so we are very limited in regards with what we can do about it.

If anyone can fix https://sourceforge.net/p/ruamel-yaml/tickets/416/ for us, I will do my best to help on allowing some level of configuration for writing.

Yep, i am planning to make a PR to address the octals issue as we not want to write (dump) yaml files that do not pass our own default rules. That is why the rewrite function is still highly-experimental and likely the place where external help could really make a dent.

**Comment by MarkusTeufelberger** [comment]
If they are not necessary, they don't need to be there. YAML has a lot of optional syntax.

**Comment by jcgruenhage** [comment]
> Why does `ansible-lint --write` remove the quotes at all? Maybe just don't delete them?

@DeadNews it doesn't remove quotes itself. The output formatter in ruamel yaml is configured in a way that doesn't consider custom yamllint rules.
